torch/csrc/distributed/c10d/NanCheck.cu

```diff
@@ -79,32 +79,66 @@ struct CheckBytePack<T, /*EltPerPack*/8> {
   }
 };
 
-// (v) Template specialization for Float8_e4m3fn.
+// (v) Template specialization for Float8 types.
 // EltPerPack = 16 / 1 = 16
 
+// We want to check 8 x FP8 simultaneously, hence this template definition.
+template<typename T>
+struct HasNanFP8x8 {
+  // I am a dumb implementation. You should never call in here, unless the check
```

Comment:
Static Assert False to raise a compile error if we call in here?

Reply:
Or at least can we issue warning somehow?

Reply:
I struggle between providing basic functionality so that user code can run without break (current code) vs speed. And eventually chose the former :)

Reply:
But your point seems better: compile error can force a developer to implement the template if they want to add a new data type to `AT_DISPATCH_FLOATING_TYPES_AND4` below.

Reply:
Hmm, it seems `static_assert` doesn't work well with template prior to c++23. I may try `=delete` as suggested [here](https://stackoverflow.com/questions/44059557/whats-the-right-way-to-call-static-assertfalse).

Reply:
@kwen2501 You could also at least unroll the loop, but the speed gains would be minimal.

Reply:
You could also use the `self != self ` to check for NaN based on the dtype, but I am not sure if that's faster / more portable.

Reply:
> @kwen2501 You could also at least unroll the loop, but the speed gains would be minimal unless the compiler realizes to inline isnan

We tried that. Since the final result is a reduction, i.e.
packHasNan = isnan(byte0) || isnan(byte1) ... || isnan(byte7),
the compiler does not seem quite willing the unroll the loop.

Reply:
@kwen2501 It doesn't like `#pragma unroll`?

Reply:
We could also manually unroll since it's only 8 elements (as painful as that would be).